Showing NPO Start Extra items

gvtulder opened the pull request:
### Functional description
<!--- Give a clear explanation of the change, fix or new feature 
that this PR contains -->
<!--- Put your text below this line -->
This adds the "Extra" items for the shows in the NPO Start channel, as listed on the [NPO Start website](https://www.npostart.nl/) and in the app.
<!--- Put your text above this line -->

### Reasoning
<!--- Provide a decent justification for this PR -->
<!--- Put your text below this line -->
Some shows have additional online-only content. See also #1649.
<!--- Put your text above this line -->

### Technical description
<!--- Please provide a technical analysis of this PR, explaining the 
 changes that were made. -->
<!--- Put your text below this line -->
This change adds an "Extra" folder to the episode list for shows that have extra content. These clips are then loaded using the NPO Start API and can be browsed and viewed using the existing codebase.

As an alternative to having them in a separate folder, the Extra clips could be added to the list of episodes with an "Extra:" prefix. This makes the pagination more complicated for shows with many episodes. I would also argue that the extra content is different from the main episodes, so it makes sense to separate them.
<!--- Put your text above this line -->

### Tasks & Activities
<!-- What other tasks or activities need to be done to complete
this PR -->

- [ ] Decide on the best way to show these clips.
- [ ] Review the changes.

basrieter commented:
Could you rebase on master to make the SonarCloud stuff work again?

gvtulder commented:
Still missing a `SONAR_TOKEN` environment variable on pull requests, apparently.

basrieter commented:
> Still missing a `SONAR_TOKEN` environment variable on pull requests, apparently.

Indeed, apparently because the PR is ran in your github fork environment, which does not have my secrets (it's a safetly thing).

basrieter commented:
Could you rebase one more time? I think I fixed it now.

basrieter commented:
I removed the other SonarCloud app. So it should work in the future now.

basrieter commented on channels/channel.nos/nos2010/chn_nos2010.py:
```diff
@@ -852,27 +852,42 @@ def extract_api_pages(self, data):
+
+            if len(extras_items) > 0:
+                Logger.debug("Adding link to extras: %s", extras_url)
+                item = FolderItem("\a.: Extra :.", extras_url, content_type=contenttype.EPISODES)
```
Would it make sense to call it "Clips" instead of "Extra's"? Then you can reuse the existing translation for that:

```python
clips = LanguageHelper.get_localized_string(LanguageHelper.Clips)
clip_title = "\a.: %s :." % (clips,)
```

Or should we make the "Extra's" a translation?

gvtulder replied:
I think that "Extra's" would be better, because that's what they are called on the website and in the app. "Clips" could also mean "Fragmenten".

basrieter replied:
Ok, then let's add "Extras" to the `resources/language/resource.language.en_gb/strings.po` and perhaps translate it in `resources/language/resource.language.nl_nl/strings.po`. This can be with ID 30368: https://github.com/retrospect-addon/plugin.video.retrospect/blob/master/resources/language/resource.language.en_gb/strings.po#L678

Then add `Extras` to https://github.com/retrospect-addon/plugin.video.retrospect/blob/master/resources/lib/helpers/languagehelper.py#L74 and assign it 30368. That way you can do:

```python 
extras = LanguageHelper.get_localized_string(LanguageHelper.Extras)
extras_title = "\a.: %s :." % (extras,)
```

gvtulder replied:
OK.

basrieter commented on channels/channel.nos/nos2010/chn_nos2010.py:
```diff
@@ -837,73 +843,101 @@ def create_npo_item(self, result_set):
+                has_fragments = True
+
+        # Obtain the POM ID for this show
+        pom = re.match(r'https://start-api.npo.nl/page/franchise/([^/?]+)',
```
Please use Regexer.do_regex()

gvtulder commented:
Two more changes:

I now also added the "Fragmenten". Especially for current affairs/talkshows, these fragments often highlight individual topics/guests, so it would be interesting to show them as well.

In the second commit, I used the `https://start-api.npo.nl/page/franchise/{pom}` to load the initial page for a show, listing the episodes and now also the Extra's and Fragmenten. However, instead of downloading the full list of all episodes, this will only show the most recent ones (the episodes listed on the first page of the website/app, basically, so that's either the most recent episodes or the last season of the series). There is a subfolder for the existing full list of episodes.
This approach has a few advantages:
* It reduces the number of API requests from at least 3 (episodes, extras, fragments) to exactly 1 (the franchise page from the API).
* It improves the load time for shows with a large number of episodes (e.g., the daily news), where the current implementation will always load several pages worth of previous year's episodes.

Of course, it does hide the older episodes one level deeper, so I'm not sure if this is your preferred solution. I added it as a separate commit, so I can easily remove it again if you prefer.

basrieter commented:
> Two more changes:
> 
> I now also added the "Fragmenten". Especially for current affairs/talkshows, these fragments often highlight individual topics/guests, so it would be interesting to show them as well.
> 
> In the second commit, I used the `https://start-api.npo.nl/page/franchise/{pom}` to load the initial page for a show, listing the episodes and now also the Extra's and Fragmenten. However, instead of downloading the full list of all episodes, this will only show the most recent ones (the episodes listed on the first page of the website/app, basically, so that's either the most recent episodes or the last season of the series). There is a subfolder for the existing full list of episodes. This approach has a few advantages:
> 
> * It reduces the number of API requests from at least 3 (episodes, extras, fragments) to exactly 1 (the franchise page from the API).
> * It improves the load time for shows with a large number of episodes (e.g., the daily news), where the current implementation will always load several pages worth of previous year's episodes.
> 
> Of course, it does hide the older episodes one level deeper, so I'm not sure if this is your preferred solution. I added it as a separate commit, so I can easily remove it again if you prefer.

Yes, I really like it! I will try to push out a new update tonight or tomorrow. I am able to use my left hand again for typing, so I am picking up some development work again.

basrieter commented:
EDIT: Perhaps update the changelog.txt?

gvtulder commented:
Yes, should be ready to merge. I merged the two commits and added a line to the changelog.txt.
